src/main/java/fi/aalto/cs/apluscourses/utils/CoursesClient.java

```diff
@@ -0,0 +1,171 @@
+package fi.aalto.cs.apluscourses.utils;
+
+import fi.aalto.cs.apluscourses.model.UnexpectedResponseException;
+
+import java.io.ByteArrayInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.net.URL;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.StatusLine;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.util.EntityUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public class CoursesClient {
+
+  /**
+   * Downloads a JSON text from the given URl and returns it in a {@link ByteArrayInputStream}.
+   * @throws IOException                 If an error (e.g. network error) occurs while downloading
+   *                                     the file.
+   * @throws UnexpectedResponseException If the response isn't as expected (i.e. a status code other
+   *                                     than 2xx or different Content-Type header).
+   */
+  @NotNull
+  public static ByteArrayInputStream fetchJson(@NotNull URL url)
+      throws IOException, UnexpectedResponseException {
+    return fetch(url, "application/json",
+        entity -> new ByteArrayInputStream(EntityUtils.toByteArray(entity)));
+  }
+
+  /**
+   * Downloads a ZIP archive from the given URL and copies it to the given file.
+   * @throws IOException                 If an error (e.g. network error) occurs while downloading
+   *                                     the file.
+   * @throws UnexpectedResponseException If the response isn't as expected (i.e. a status code other
+   *                                     than 2xx or different Content-Type header).
+   */
+  public static void fetchZip(@NotNull URL url, @NotNull File file)
+      throws IOException, UnexpectedResponseException {
+    fetch(url, "application/zip", entity -> {
+      FileUtils.copyInputStreamToFile(entity.getContent(), file);
+      return null; // The return value gets ignored here
```

Comment:
Nit: This looks a bit like abuse of mapper interface, as a programmer would usually expect mapping to be free of side-effects. Here, the mapper is more like a consumer, right? But fair enough. :)

Reply:
Yes, you are 100% correct, this was laziness on my part :smile: I'll probably add a consumer functional interface